### PR TITLE
4655 key not found exception

### LINKED ISSUE #4655: KeyNotFound Exception in ProcedureCanBeWrittenAsFunctionInspection
Discovered in the logs for #4644:

```
2018-12-17 14:27:44.0304;WARN-2.3.1.4308;Rubberduck.Inspections.Rubberduck.Inspections.Inspector;System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Rubberduck.Inspections.Concrete.ProcedureCanBeWrittenAsFunctionInspection.<>c__DisplayClass4_0.<DoGetInspectionResults>b__3(QualifiedContext`1 context) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Concrete\ProcedureCanBeWrittenAsFunctionInspection.cs:line 40
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at Rubberduck.Inspections.Abstract.InspectionBase.GetInspectionResults(CancellationToken token) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Abstract\InspectionBase.cs:line 168
   at Rubberduck.Inspections.Rubberduck.Inspections.Inspector.RunInspection(IInspection inspection, ConcurrentBag`1 allIssues, CancellationToken token) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Inspector.cs:line 188;System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Rubberduck.Inspections.Concrete.ProcedureCanBeWrittenAsFunctionInspection.<>c__DisplayClass4_0.<DoGetInspectionResults>b__3(QualifiedContext`1 context) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Concrete\ProcedureCanBeWrittenAsFunctionInspection.cs:line 40
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   at Rubberduck.Inspections.Abstract.InspectionBase.GetInspectionResults(CancellationToken token) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Abstract\InspectionBase.cs:line 168
   at Rubberduck.Inspections.Rubberduck.Inspections.Inspector.RunInspection(IInspection inspection, ConcurrentBag`1 allIssues, CancellationToken token) in C:\projects\rubberduck\Rubberduck.CodeAnalysis\Inspections\Inspector.cs:line 188
```

### PR DESCRIPTION
**Pull request by BZngr**
Closes #4655 

Added `TryGetValue()` where the inspection accessed a dictionary using the key without verification checks.  Since I could not determine a specific use case that caused the missing key issue, adding protection to accessing the dictionary may be all that is/was needed to clear the issue.

### CODE REVIEW

**Comment by BZngr** [comment]
While trying to find a use case that would result in the `KeyNotFoundException`, discovered that the inspection does a lot of processing prior to checking if their are any valid contexts to evaluate.  So...added a check at the top of the inspection.

File: Rubberduck.CodeAnalysis/Inspections/Concrete/ProcedureCanBeWrittenAsFunctionInspection.cs [hunk 1]
```diff
@@ -36,14 +41,22 @@ protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
                 .Concat(builtinHandlers)
                 .Concat(userDeclarations.Where(item => item.IsWithEvents)));
 
+            bool HasArgumentReferencesWithIsAssignmentFlagged(QualifiedContext<ParserRuleContext> context)
+                => contextLookup.TryGetValue(context.Context.GetChild<VBAParser.ArgContext>(), out Declaration decl)
+                    ? decl.References.Any(rf => rf.IsAssignment) : false;
```
Comment (MDoerner) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ProcedureCanBeWrittenAsFunctionInspection.cs]: The `out` variable could be implicily types with `var`.

Comment (MDoerner) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ProcedureCanBeWrittenAsFunctionInspection.cs]: I am not too sure whether a local function is the most readable thing here, especially an expression bodied one. Moreover, I would put each part of the ternary on a separate line. 
However, both might be personal taste. (The local function might be less surpising when not expression bodied.)

File: Rubberduck.CodeAnalysis/Inspections/Concrete/ProcedureCanBeWrittenAsFunctionInspection.cs [hunk 2]
```diff
@@ -36,14 +41,22 @@ protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
...
+
+            Declaration GetSubStmtParentDeclaration(QualifiedContext<ParserRuleContext> context)
+                => contextLookup.TryGetValue((VBAParser.SubStmtContext)context.Context.Parent, out Declaration decl)
+                    ? decl : null;
```
Comment (MDoerner) [inline Rubberduck.CodeAnalysis/Inspections/Concrete/ProcedureCanBeWrittenAsFunctionInspection.cs]: dito

**Pull request merged**